src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,348 @@
+        std::cout << "Calculating bondstress for Case Ic"
+                  << ".\n";
+      _bond_stress = -1.9 * _max_bondstress;
+    }
+    else
+    {
+      if (_debug)
+        // if (_current_node->id() == 144)
+        std::cout << "Calculating bondstress for Case Id"
+                  << ".\n";
+      _bond_stress = _frictional_bondstress * MathUtils::sign(slip);
+    }
+  }
+  else if (slip > plastic_slip_max && slip < bond_slip.slip_max)
+  {
+    if (_debug)
+      // if (_current_node->id() == 144)
+      std::cout << "Calculating bondstress for Case II"
+                << ".\n";
+
+    _bond_stress = (slip - plastic_slip_max) * bond_slip.bondstress_max /
+                   (bond_slip.slip_max - plastic_slip_max);
+  }
+  else if (slip < plastic_slip_min && slip > bond_slip.slip_min)
+  {
+    if (_debug)
+      // if (_current_node->id() == 144)
+      std::cout << "Calculating bondstress for Case III"
+                << ".\n";
+
+    _bond_stress = (slip - plastic_slip_min) * bond_slip.bondstress_min /
+                   (bond_slip.slip_min - plastic_slip_min);
+  }
+  else
+    _bond_stress = _frictional_bondstress;
+
+  if (_debug)
+    // if (_current_node->id() == 144)
+    std::cout << "Bondstress = " << _bond_stress << "\n";
+
+  Real bond_force = 2.0 * libMesh::pi * _bar_radius * _current_elem_volume * _bond_stress;
+
+  RealVectorValue constraint_force_axial = bond_force * _slave_tangent;
+  RealVectorValue constraint_force_normal = _penalty * slip_normal;
+
+  _constraint_residual = constraint_force_axial + constraint_force_normal;
+
+  if (_debug)
+  // if (_current_node->id() == 144)
+  {
+    std::cout << "Constraint Residual Axial = " << constraint_force_axial << "\n";
+    std::cout << "Constraint Residual Normal = " << constraint_force_normal << "\n";
+    std::cout << "Constraint Residual = " << _constraint_residual << "\n";
+  }
+
+  bond_slip.bondstress_min = std::min(bond_slip.bondstress_min_old, _bond_stress);
+  bond_slip.bondstress_max = std::max(bond_slip.bondstress_max_old, _bond_stress);
+
+  if (_fe_problem.converged())
+  {
+    _bondslip[node->id()].slip_min = bond_slip.slip_min;
+    _bondslip[node->id()].slip_max = bond_slip.slip_max;
+    _bondslip[node->id()].bondstress_min = bond_slip.bondstress_min;
+    _bondslip[node->id()].bondstress_max = bond_slip.bondstress_max;
+  }
+}
+
+Real
+RebarBondSlipConstraint::computeQpResidual(Moose::ConstraintType type)
+{
+  Real resid = _constraint_residual(_component);
+
+  switch (type)
+  {
+    case Moose::Slave:
+      return _test_slave[_i][_qp] * resid;
+
+    case Moose::Master:
+      return _test_master[_i][_qp] * -resid;
+  }
+
+  return 0.0;
+}
+
+Real
+RebarBondSlipConstraint::computeQpJacobian(Moose::ConstraintJacobianType type)
+{
+  switch (type)
+  {
+    case Moose::SlaveSlave:
+      return _phi_slave[_j][_qp] * _penalty * _test_slave[_i][_qp] *
+             (1.0 - _slave_tangent(_component) * _slave_tangent(_component));
```

Comment:
These are all computed assuming that there is a penalty stiffness in the axial and perpendicular directions, which isn't the case anymore with the bond slip model. You would need to apply the penalty stiffness in the perpendicular directions, but subtract that off in the axial direction and replace it with the tangent from the bond slip model.  We do that kind of thing in contact in the MechanicalContactConstraint class.  Same idea here. I imagine this is the main source of the convergence trouble.

Reply:
I was just thinking -- there's a newly introduced Constraint in tensor_mechanics that Swetha wrote that does something very similar to this Constraint: NodalFrictionalConstraint.  It represents a frictional slider that uses a penalty constraint until it slips. I just looked at the code for the Jacobian in that class. It computes the nonlinear Jacobian, but it doesn't have any of the code I was expecting to see to deal with the directionality of the Jacobian.  That's because the NodalFrictionalConstraint class is only applied to the dof in the axial direction of the beam, and it assumes that the beam is aligned in one of the Cartesian directions. Another separate Constraint (NodalStickConstraint) deals with tying it in the other two directions. Your constraint is dealing with all directions (which is probably preferable).

Reply:
The bondslip contribution towards the residual is constant. So I only considered the penalty contributions towards jacobian calculation. I believe I am using the normal direction to assign the contribution. If the truss member is inclined, only then the penalty would apply to  both x- and y- directions.